pkg/testing/fetcher_artifact.go

```diff
@@ -138,7 +138,7 @@ func (r *artifactResult) Fetch(ctx context.Context, l Logger, dir string) error
 
 func findURI(ctx context.Context, doer httpDoer, version *semver.ParsedSemVer) (string, error) {
 	// if we know the exact build ID, we can build the URI right away
-	if version.BuildMetadata() != "" {
+	if version.BuildMetadata() != "" && !version.IsIndependentRelease() {
```

Comment:
i think these are 2 different things, `a1fab` in `8.16.0-a1fab` should be captured as `prerelease` named group not `buildmetadata`
build metadata are prefixed with `+`

Reply:
maybe this should check for prerelease and metadata (?)

Reply:
> @michalpristas the way the snapshot API works it always requires us to provide a build ID as a part of the URL. We support 3 different kinds of versions here:
> 
>     1. This particular condition means that the `semver.ParsedSemVer` passed to this function already contains this build ID (which we store in `version.BuildMetadata()`, e.g. `8.16.0-a1fab`) and we don't need to perform any of additional requests to find it. Hence, we just return the composed string right away. When this code was written we supported versions with build IDs in them, I'm not 100% sure it's true today. Perhaps @pchila knows.
> 

This is the infamous "upgrade to specific snapshot" feature which by now we discussed many times: 
- the version string `x.y.z-SNAPSHOT+<buildid>` identifies a specific build for a snapshot version and it's used by the upgrade command to directly fetch a snapshot build (in elastic-agent only the SNAPSHOT downloader will check for this format, so this can be applied only for snapshot builds)
- the version string `x.y.z+build<long date string>` is an independent agent release (as of today not supported for snapshots) and that has a completely different meaning (it's a re-release of last released version with some patches) and it should not be handled in the same way

I didn't have a look at the PR yet, just responding to the ping, I will have another look here to understand why we need to do this sort of check (without checking the rest of the code I tend to agree with @michalpristas that we need to add a restriction on this being a snapshot build)

Reply:
ok so condition should check for both build meta and is snapshot, i'll update it.